src/test/java/com/amplitude/AmplitudeTest.java

```diff
@@ -335,6 +341,51 @@ public void testWithSwallowMiddleware()
     assertEquals("middleware_user_2", sentEvent.userId);
   }
 
+  @Test
+  public void testShouldWait()
+      throws NoSuchFieldException, IllegalAccessException, AmplitudeInvalidAPIKeyException,
+          InterruptedException, BrokenBarrierException {
+    Amplitude amplitude = Amplitude.getInstance();
+    amplitude.init(apiKey);
+    Event event = new Event("test event", "test-user-0");
+    Event event2 = new Event("test event", "test-user-1");
+    HttpCall httpCall = getMockHttpCall(amplitude, false);
+    CyclicBarrier barrier = new CyclicBarrier(2);
+    CountDownLatch latch = new CountDownLatch(1);
+    Response response = new Response();
+    response.status = Status.RATELIMIT;
+    response.code = 429;
+    response.rateLimitBody = new JSONObject();
+    response.rateLimitBody.put("throttledUsers", new JSONObject());
+    response.rateLimitBody.put("throttledDevices", new JSONObject());
+    response.rateLimitBody.put("exceededDailyQuotaDevices", new JSONObject());
+    response.rateLimitBody.put("exceededDailyQuotaUsers", new JSONObject());
+    response.rateLimitBody.getJSONObject("throttledUsers").put("test-user-0", 15);
+    response.rateLimitBody.put("throttledEvents", new int[] {0});
+    when(httpCall.makeRequest(anyList()))
+        .thenAnswer(
+            invocation -> {
+              return response;
+            })
+        .thenAnswer(
+            invocation -> {
+              latch.countDown();
+              barrier.await();
+              return response;
+            });
+    assertFalse(amplitude.shouldWait(event));
+    Field httpTransportField = amplitude.getClass().getDeclaredField("httpTransport");
+    httpTransportField.setAccessible(true);
+    HttpTransport transport = (HttpTransport) httpTransportField.get(amplitude);
+    amplitude.setRecordThrottledId(true);
+    assertFalse(amplitude.shouldWait(event));
+    transport.retryEvents(Arrays.asList(event, event2), response);
+    assertTrue(latch.await(1, TimeUnit.SECONDS));
+    assertTrue(amplitude.shouldWait(event));
+    assertFalse(amplitude.shouldWait(event2));
+    barrier.await();
```

Comment:
Nit: anything we can think of after do `barrier.await()`? Since we release the barrier now, maybe in callback we can assert the shouldWait to be false.